### ISSUE TITLE
File based sessions do not work as they should

### ISSUE THREAD

**Issue by davidalger** [davidalger/warden#36]
Although it is recommended to use redis (and used in the example install) for session storage, file based sessions should work. The `/var/lib/php/fpm/session` directory does not exist in the fpm images, and the `/var/lib/php/fpm` appears to have a wrong group. These two issues need to be corrected, and then this should resolve itself.

```
1 exception(s):
Exception #0 (Exception): Warning: SessionHandler::read(): open(/var/lib/php/fpm/session/sess_o02fipeq95citdo4dr83qlj67h, O_RDWR) failed: Permission denied (13) in /var/www/html/vendor/magento/framework/Session/SaveHandler/Native.php on line 22

Exception #0 (Exception): Warning: SessionHandler::read(): open(/var/lib/php/fpm/session/sess_o02fipeq95citdo4dr83qlj67h, O_RDWR) failed: Permission denied (13) in /var/www/html/vendor/magento/framework/Session/SaveHandler/Native.php on line 22
#0 [internal function]: Magento\Framework\App\ErrorHandler->handler(2, 'SessionHandler:...', '/var/www/html/v...', 22, Array)
#1 /var/www/html/vendor/magento/framework/Session/SaveHandler/Native.php(22): SessionHandler->read('o02fipeq95citdo...')
#2 /var/www/html/vendor/magento/framework/Session/SaveHandler.php(80): Magento\Framework\Session\SaveHandler\Native->read('o02fipeq95citdo...')
#3 [internal function]: Magento\Framework\Session\SaveHandler->read('o02fipeq95citdo...')
#4 /var/www/html/vendor/magento/framework/Session/SessionManager.php(191): session_start()
#5 /var/www/html/vendor/magento/framework/Interception/Interceptor.php(58): Magento\Framework\Session\SessionManager->start()
#6 /var/www/html/vendor/magento/framework/Interception/Interceptor.php(138): Magento\Framework\Session\Generic\Interceptor->___callParent('start', Array)
#7 /var/www/html/vendor/magento/framework/Interception/Interceptor.php(153): Magento\Framework\Session\Generic\Interceptor->Magento\Framework\Interception\{closure}()
#8 /var/www/html/generated/code/Magento/Framework/Session/Generic/Interceptor.php(52): Magento\Framework\Session\Generic\Interceptor->___callPlugins('start', Array, Array)
#9 /var/www/html/vendor/magento/framework/Session/SessionManager.php(127): Magento\Framework\Session\Generic\Interceptor->start()
#10 /var/www/html/generated/code/Magento/Framework/Session/Generic/Interceptor.php(14): Magento\Framework\Session\SessionManager->__construct(Object(Magento\Framework\App\Request\Http), Object(Magento\Framework\Session\SidResolver\Proxy), Object(Magento\Framework\Session\Config), Object(Magento\Framework\Session\SaveHandler), Object(Magento\Framework\Session\Validator), Object(Magento\Framework\Session\Storage), Object(Magento\Framework\Stdlib\Cookie\PhpCookieManager), Object(Magento\Framework\Stdlib\Cookie\CookieMetadataFactory), Object(Magento\Framework\App\State\Interceptor))
#11 /var/www/html/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(116): Magento\Framework\Session\Generic\Interceptor->__construct(Object(Magento\Framework\App\Request\Http), Object(Magento\Framework\Session\SidResolver\Proxy), Object(Magento\Framework\Session\Config), Object(Magento\Framework\Session\SaveHandler), Object(Magento\Framework\Session\Validator), Object(Magento\Framework\Session\Storage), Object(Magento\Framework\Stdlib\Cookie\PhpCookieManager), Object(Magento\Framework\Stdlib\Cookie\CookieMetadataFactory), Object(Magento\Framework\App\State\Interceptor))
#12 /var/www/html/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(66): Magento\Framework\ObjectManager\Factory\AbstractFactory->createObject('Magento\\Framewo...', Array)
#13 /var/www/html/vendor/magento/framework/ObjectManager/ObjectManager.php(70): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create('Magento\\Framewo...')
#14 /var/www/html/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(160): Magento\Framework\ObjectManager\ObjectManager->get('Magento\\Framewo...')
#15 /var/www/html/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(246): Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgument(Array, 'Magento\\Framewo...', NULL, 'session', 'Magento\\Framewo...')
#16 /var/www/html/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(34): Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgumentsInRuntime('Magento\\Framewo...', Array, Array)
#17 /var/www/html/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(59): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->_resolveArguments('Magento\\Framewo...', Array, Array)
#18 /var/www/html/vendor/magento/framework/ObjectManager/ObjectManager.php(70): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create('Magento\\Framewo...')
#19 /var/www/html/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(160): Magento\Framework\ObjectManager\ObjectManager->get('Magento\\Framewo...')
#20 /var/www/html/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(246): Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgument(Array, 'Magento\\Framewo...', NULL, 'formKey', 'Magento\\Framewo...')
#21 /var/www/html/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(34): Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgumentsInRuntime('Magento\\Framewo...', Array, Array)
#22 /var/www/html/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(59): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->_resolveArguments('Magento\\Framewo...', Array, Array)
#23 /var/www/html/vendor/magento/framework/ObjectManager/ObjectManager.php(70): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create('Magento\\Framewo...')
#24 /var/www/html/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(160): Magento\Framework\ObjectManager\ObjectManager->get('Magento\\Framewo...')
#25 /var/www/html/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(246): Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgument(Array, 'Magento\\Framewo...', NULL, 'formKeyValidato...', 'CsrfRequestVali...')
#26 /var/www/html/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(34): Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgumentsInRuntime('CsrfRequestVali...', Array, Array)
#27 /var/www/html/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(59): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->_resolveArguments('CsrfRequestVali...', Array, Array)
#28 /var/www/html/vendor/magento/framework/ObjectManager/ObjectManager.php(70): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create('CsrfRequestVali...')
#29 /var/www/html/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(196): Magento\Framework\ObjectManager\ObjectManager->get('CsrfRequestVali...')
#30 /var/www/html/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(172): Magento\Framework\ObjectManager\Factory\AbstractFactory->parseArray(Array)
#31 /var/www/html/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(246): Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgument(Array, NULL, NULL, 'validators', 'RequestValidato...')
#32 /var/www/html/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(34): Magento\Framework\ObjectManager\Factory\AbstractFactory->resolveArgumentsInRuntime('RequestValidato...', Array, Array)
#33 /var/www/html/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(59): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->_resolveArguments('RequestValidato...', Array, Array)
#34 /var/www/html/vendor/magento/framework/ObjectManager/ObjectManager.php(70): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create('RequestValidato...')
#35 /var/www/html/vendor/magento/framework/App/FrontController.php(71): Magento\Framework\ObjectManager\ObjectManager->get('RequestValidato...')
#36 /var/www/html/generated/code/Magento/Framework/App/FrontController/Interceptor.php(14): Magento\Framework\App\FrontController->__construct(Object(Magento\Framework\App\RouterList), Object(Magento\Framework\App\Response\Http\Interceptor), NULL, NULL, NULL)
#37 /var/www/html/vendor/magento/framework/ObjectManager/Factory/AbstractFactory.php(116): Magento\Framework\App\FrontController\Interceptor->__construct(Object(Magento\Framework\App\RouterList), Object(Magento\Framework\App\Response\Http\Interceptor), NULL, NULL, NULL)
#38 /var/www/html/vendor/magento/framework/ObjectManager/Factory/Dynamic/Developer.php(66): Magento\Framework\ObjectManager\Factory\AbstractFactory->createObject('Magento\\Framewo...', Array)
#39 /var/www/html/vendor/magento/framework/ObjectManager/ObjectManager.php(70): Magento\Framework\ObjectManager\Factory\Dynamic\Developer->create('Magento\\Framewo...')
#40 /var/www/html/vendor/magento/framework/App/Http.php(134): Magento\Framework\ObjectManager\ObjectManager->get('Magento\\Framewo...')
#41 /var/www/html/generated/code/Magento/Framework/App/Http/Interceptor.php(24): Magento\Framework\App\Http->launch()
#42 /var/www/html/vendor/magento/framework/App/Bootstrap.php(258): Magento\Framework\App\Http\Interceptor->launch()
#43 /var/www/html/pub/index.php(37): Magento\Framework\App\Bootstrap->run(Object(Magento\Framework\App\Http\Interceptor))
#44 {main}
```

**Comment by davidalger** [comment]
Dug into this a little bit more. This is due to running the container as `www-data` user and the pkg install owning/grouping a bunch of things with default `php-fpm` user, which normally if container was executed as root, would work. Notice the difference when the container is run as root: the php-fpm workers are running as php-fpm user vs www-data user. That would be good in production, but not in a local dev environment. :) 

```
davidalger:04:04 PM:~$ docker run --rm -d --name test-01 davidalger/warden:mage2-fpm-7.2
77e5a0cb324dfada7e3bb519ca9496480b85875c05ee7ce6b88bb08042a35a43
davidalger:04:05 PM:~$ docker exec -it test-01 bash
www-data@77e5a0cb324d:10:05 PM:/var/www/html$ grep 'group = ' /etc/php-fpm.d/www.conf 
group = php-fpm
;listen.group = root
www-data@77e5a0cb324d:10:05 PM:/var/www/html$ ps uax
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
www-data     1  0.3  0.1 832776 35292 ?        Ss   22:05   0:00 php-fpm: master process (/etc/php-fpm.conf)
www-data    14  0.0  0.0 832776  7636 ?        S    22:05   0:00 php-fpm: pool www
www-data    15  0.0  0.0 832776  7636 ?        S    22:05   0:00 php-fpm: pool www
www-data    16  0.0  0.0 832776  7636 ?        S    22:05   0:00 php-fpm: pool www
www-data    17  0.0  0.0 832776  7640 ?        S    22:05   0:00 php-fpm: pool www
www-data    18  0.0  0.0 832776  7640 ?        S    22:05   0:00 php-fpm: pool www
www-data    19  0.3  0.0  12764  4000 pts/0    Ss   22:05   0:00 bash
www-data    44  0.0  0.0  51752  3416 pts/0    R+   22:05   0:00 ps uax
www-data@77e5a0cb324d:10:05 PM:/var/www/html$ sudo -i
root@77e5a0cb324d:10:05 PM:~# ls -lhd /var/lib/php/fpm/*
drwxrwx--- 2 root php-fpm 4.0K Oct 24 19:58 /var/lib/php/fpm/opcache
drwxrwx--- 2 root php-fpm 4.0K Oct 24 19:58 /var/lib/php/fpm/session
drwxrwx--- 2 root php-fpm 4.0K Oct 24 19:58 /var/lib/php/fpm/wsdlcache
root@77e5a0cb324d:10:05 PM:~# logout
www-data@77e5a0cb324d:10:05 PM:/var/www/html$ exit
davidalger:04:05 PM:~$ docker kill test-01
test-01
davidalger:04:05 PM:~$ docker run --rm -d --name test-01 --user root  davidalger/warden:mage2-fpm-7.2
c3a0cab469f4019df8ef5567f8c3c2473da0b736595620e24d197039dcf79652
davidalger:04:06 PM:~$ docker exec -it test-01 bash
root@c3a0cab469f4:10:06 PM:/var/www/html# ls -lhd /var/lib/php/fpm/*
drwxrwx--- 2 root php-fpm 4.0K Oct 24 19:58 /var/lib/php/fpm/opcache
drwxrwx--- 2 root php-fpm 4.0K Oct 24 19:58 /var/lib/php/fpm/session
drwxrwx--- 2 root php-fpm 4.0K Oct 24 19:58 /var/lib/php/fpm/wsdlcache
root@c3a0cab469f4:10:06 PM:/var/www/html# ps uax
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root         1  0.4  0.1 832776 35444 ?        Ss   22:06   0:00 php-fpm: master process (/etc/php-fpm.conf)
php-fpm     13  0.0  0.0 832776 12728 ?        S    22:06   0:00 php-fpm: pool www
php-fpm     14  0.0  0.0 832776 12728 ?        S    22:06   0:00 php-fpm: pool www
php-fpm     15  0.0  0.0 832776 12728 ?        S    22:06   0:00 php-fpm: pool www
php-fpm     16  0.0  0.0 832776 12732 ?        S    22:06   0:00 php-fpm: pool www
php-fpm     17  0.0  0.0 832776 12736 ?        S    22:06   0:00 php-fpm: pool www
root        18  0.5  0.0  12880  4032 pts/0    Ss   22:06   0:00 bash
root        41  0.0  0.0  51752  3480 pts/0    R+   22:06   0:00 ps uax
root@c3a0cab469f4:10:06 PM:/var/www/html# exit
```

Going to work on a couple updates to the startup script in the base fpm image to own/group any files owned/grouped as `php-fpm` to the user container is started as if the user is not root.

**Comment by davidalger** [comment]
File based sessions should now work as expected. The resolution here was to remove setup of `www-data` from base (generic) `fpm` images (making them more portable as a base really, and usable on production setups in the future) and move that into the Warden specific `fpm` images. Once I did that, fixing this was as simple and as complicated as resolving the owner/group on files in `/var/log` and `/var/lib` which need to be owned/grouped by the user which `php-fpm` workers run under.

Big thank you to @martinssu, @adexandros, and @vbuck for bringing this issue to my attention. I've tested this on a local fresh install of Magento 2.3.3, but let me know if you guys see this again (although I would still recommend using redis sessions as a matter of std practice, even on local).

To fix the issue on local environments, simply run the following command to pull updated images and recreate affected containers:

```
warden env pull && warden env up -d && warden sync start
```
